job/cache.go

```diff
@@ -288,3 +296,49 @@ func (c *LockFreeJobCache) PersistEvery(persistWaitTime time.Duration) {
 		}
 	}
 }
+
+func (c *LockFreeJobCache) locateJobStatsIndexForRetention(stats []*JobStat) (marker int) {
+	now := time.Now()
+	expiresAt := now.Add(-c.retentionPeriod)
+	pos := -1
+	for i, el := range stats {
+		diff := el.RanAt.Sub(expiresAt)
+		if diff < 0 {
+			pos = i
+		}
+	}
+	return pos
+}
+
+func (c *LockFreeJobCache) runCompactJobStatsCycle() error {
+	for el := range c.jobs.Iter() {
+		job := (*Job)(el.Value)
+		c.compactJobStats(job)
+	}
+	return nil
+}
+
+func (c *LockFreeJobCache) compactJobStats(job *Job) error {
+	job.lock.RLock()
+	defer job.lock.RUnlock()
+	pos := c.locateJobStatsIndexForRetention(job.Stats)
+	if pos >= 0 {
+		log.Infof("JobStats TTL: removing %d items", pos+1)
+		tmp := make([]*JobStat, len(job.Stats)-pos-1)
+		copy(tmp, job.Stats[pos+1:])
+		job.Stats = tmp
+	}
+	return nil
+}
+
+func (c *LockFreeJobCache) RetentionEvery(retentionWaitTime time.Duration) {
```

Comment:
@QBui Could you create a test case for this function? Also, Do you think it should be called "RetainEvery" to better match "PersistEvery" since retain and persist are both verbs?